File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0373_Find_K_Pairs_with_Smallest_Sums.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;

public class Question_0373_Find_K_Pairs_with_Smallest_Sums {
    public List<List<Integer>> kSmallestPairs(int[] nums1, int[] nums2, int k) {
        List<List<Integer>> res = new ArrayList<>();
        if (nums1 == null || nums1.length == 0 || nums2 == null || nums2.length == 0 || k <= 0) {
            return res;
        }
        PriorityQueue<int[]> queue = new PriorityQueue<>((a, b) -> a[0] + a[1] - b[0] - b[1]);
        for (int i = 0; i < nums1.length && i < k; i++) {
            queue.offer(new int[]{nums1[i], nums2[0], 0});
        }
        while (k-- > 0 && !queue.isEmpty()) {
            int[] cur = queue.poll();
            List<Integer> list = new ArrayList<>();
            list.add(cur[0]);
            list.add(cur[1]);
            res.add(list);
            if (cur[2] == nums2.length - 1) {
                continue;
            }
            queue.offer(new int[]{cur[0], nums2[cur[2] + 1], cur[2] + 1});
        }
        return res;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0373_Find_K_Pairs_with_Smallest_Sums.cc
#include <vector>
#include <queue>

using namespace std;

class Solution {
 public:
  vector<vector<int>> kSmallestPairs(vector<int>& nums1, vector<int>& nums2, int k) {
    vector<vector<int>> res;
    if (nums1.empty() || nums2.empty() || k <= 0) {
      return res;
    }
    auto comp = [](vector<int> a, vector<int> b) {
        return a[0] + a[1] > b[0] + b[1];
    };
    priority_queue<vector<int>, vector<vector<int>>, decltype(comp)> q(comp);
    for (int i = 0; i < nums1.size(); i++) {
      q.emplace(vector<int>{nums1[i], nums2[0], 0});
    }
    while (k-- > 0 && !q.empty()) {
      vector<int> vec = q.top();
      q.pop();
      res.push_back({vec[0], vec[1]});
      if (vec[2] == nums2.size() - 1) {
          continue;
      }
      q.emplace(vector<int>{vec[0], nums2[vec[2] + 1], vec[2] + 1});
    }
    return res;
  }
};
